pytensor/link/jax/ops.py

```diff
@@ -0,0 +1,424 @@
+        )
+        vars_from_func = tree_map(lambda x: x.get_vars(), func_vars)
+        pt_vars = dict(vars=pt_vars, vars_from_func=vars_from_func)
+        """
+        def func_unwrapped(vars_all, static_vars):
+            vars, vars_from_func = vars_all["vars"], vars_all["vars_from_func"]
+            func_vars_evaled = tree_map(
+                lambda x, y: x.get_func_with_vars(y), func_vars, vars_from_func
+            )
+            args, kwargs = eqx.combine(vars, static_vars, func_vars_evaled)
+            return self.jaxfunc(*args, **kwargs)
+        """
+
+        pt_vars_flat, vars_treedef = tree_flatten(pt_vars)
+        pt_vars_types_flat = [var.type for var in pt_vars_flat]
+        shapes_vars_flat = pytensor.compile.builders.infer_shape(pt_vars_flat, (), ())
+        shapes_vars = tree_unflatten(vars_treedef, shapes_vars_flat)
+
+        dummy_inputs_jax = jax.tree_util.tree_map(
+            lambda var, shape: jnp.empty(
+                [int(dim.eval()) for dim in shape], dtype=var.type.dtype
+            ),
+            pt_vars,
+            shapes_vars,
+        )
+
+        # Combine the static variables with the inputs, and split them again in the
+        # output. Static variables don't take part in the graph, or might be a
+        # a function that is returned.
+        jaxfunc_partitioned, static_out_dic = _partition_jaxfunc(
+            jaxfunc, static_vars, func_vars
+        )
+
+        func_flattened = _flatten_func(jaxfunc_partitioned, vars_treedef)
+
+        jaxtypes_outvars = jax.eval_shape(
+            ft.partial(jaxfunc_partitioned, vars=dummy_inputs_jax),
+        )
+
+        jaxtypes_outvars_flat, outvars_treedef = tree_flatten(jaxtypes_outvars)
+
+        pttypes_outvars = [
+            pt.TensorType(dtype=var.dtype, shape=var.shape)
+            for var in jaxtypes_outvars_flat
+        ]
+
+        ### Call the function that accepts flat inputs, which in turn calls the one that
+        ### combines the inputs and static variables.
+        jitted_sol_op_jax = jax.jit(func_flattened)
+        len_gz = len(pttypes_outvars)
+
+        vjp_sol_op_jax = _get_vjp_sol_op_jax(func_flattened, len_gz)
+        jitted_vjp_sol_op_jax = jax.jit(vjp_sol_op_jax)
+
+        if name is None:
+            curr_name = jaxfunc.__name__
+        else:
+            curr_name = name
+
+        # Get classes that creates a Pytensor Op out of our function that accept
+        # flattened inputs. They are created each time, to set a custom name for the
+        # class.
+        SolOp, VJPSolOp = _return_pytensor_ops_classes(curr_name)
+
+        local_op = SolOp(
+            vars_treedef,
+            outvars_treedef,
+            input_types=pt_vars_types_flat,
+            output_types=pttypes_outvars,
+            jitted_sol_op_jax=jitted_sol_op_jax,
+            jitted_vjp_sol_op_jax=jitted_vjp_sol_op_jax,
+        )
+
+        @jax_funcify.register(SolOp)
```

Comment:
I guess we can dispatch on the base class just once?

Reply:
I mean you can define SolOp class outside the decorator and dispatch on that.

Then the decorator can return a subclass of that and you don't need to bother dispatching because the base class dispatch will cover it